src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs

```diff
@@ -439,6 +439,37 @@ private void FailActivationUnsynchronized(HttpContext nonClonedContext, ILogger
             }
         }
 
+        internal async Task<bool> CancelPreviousPoll(HttpContext context)
+        {
+            CancellationTokenSource cts;
+            lock (_stateLock)
+            {
+                // Need to sync cts access with DisposeAsync as that will dispose the cts
+                cts = Status == HttpConnectionStatus.Disposed ? null : Cancellation;
+            }
+
+            using (cts)
+            {
+                // Cancel the previous request
+                cts?.Cancel();
```

Comment:
If you don't keep _stateLock trough the call to Cancel, this could still ODE as this gets disposed by another thread calling DispoasAsync(). After the _stateLock is released by CancelPreviousPoll(), DispoasAsync() can acquire the lock, see the HttpConnectionStatus still isn't dispose and start disposing the Cancellation CTS before CancelPreviousPoll() ever calls Cancel() because the lock was released.

Reply:
Whoops, I knew that and meant to null it out here...

Reply:
Cancellation can never get reinitialized, right?

Reply:
It can, but it's in `TryActivateLongPollingConnection` which is inside the lock

Reply:
Make sure `TryActivateLongPollingConnection` cannot recreate the CTS. It doesn't look like TryActivateLongPollingConnection checks whether Status is HttpConnectionStatus.Disposed currently.

Reply:
We both didn't line up the curly braces correctly when looking at this :D
If the status != inactive it goes to the `FailActivationUnsynchronized` call which will set the correct response and return